// ftp-srv/src/commands/index.js
const _ = require('lodash');
const Promise = require('bluebird');
const {TextDecoder} = require('util');

const REGISTRY = require('./registry');

const CMD_FLAG_REGEX = new RegExp(/^-(\w{1})$/);

function scoreDecodedArg(text) {
  const raw = String(text || '');
  if (!raw) return -Infinity;
  let score = 0;
  score += (raw.match(/[\u4E00-\u9FFF]/g) || []).length * 8;
  score += (raw.match(/[A-Za-z0-9._/\\:-]/g) || []).length * 2;
  score -= (raw.match(/[\uFFFD]/g) || []).length * 12;
  score -= (raw.match(/[^\u4E00-\u9FFFA-Za-z0-9._/\\:\-]/g) || []).length * 2;
  return score;
}

function decodeCommandArg(text) {
  const latin1 = String(text || '');
  if (!latin1) return latin1;
  const raw = Buffer.from(latin1, 'latin1');
  const candidates = [];
  try {
    candidates.push(new TextDecoder('utf-8', {fatal: false}).decode(raw));
  } catch {}
  try {
    candidates.push(new TextDecoder('gb18030', {fatal: false}).decode(raw));
  } catch {}
  candidates.push(latin1);
  let best = latin1;
  let bestScore = scoreDecodedArg(best);
  for (const candidate of candidates) {
    const score = scoreDecodedArg(candidate);
    if (score > bestScore) {
      best = candidate;
      bestScore = score;
    }
  }
  return best;
}

class FtpCommands {
  constructor(connection) {
    this.connection = connection;
    this.previousCommand = {};
    this.blacklist = _.get(this.connection, 'server.options.blacklist', []).map((cmd) => _.upperCase(cmd));
    this.whitelist = _.get(this.connection, 'server.options.whitelist', []).map((cmd) => _.upperCase(cmd));
  }

  parse(message) {
    const strippedMessage = message.replace(/"/g, '');
    let [directive, ...args] = strippedMessage.split(' ');
    directive = _.chain(directive).trim().toUpper().value();

    const parseCommandFlags = !['RETR', 'SIZE', 'STOR'].includes(directive);
    const params = args.reduce(({arg, flags}, param) => {
      if (parseCommandFlags && CMD_FLAG_REGEX.test(param)) flags.push(param);
      else arg.push(param);
      return {arg, flags};
    }, {arg: [], flags: []});

    const command = {
      directive,
      arg: params.arg.length ? decodeCommandArg(params.arg.join(' ')) : null,
      flags: params.flags,
      raw: message
    };
    return command;
  }

  handle(command) {
    if (typeof command === 'string') command = this.parse(command);

    // Obfuscate password from logs
    const logCommand = _.clone(command);
    if (logCommand.directive === 'PASS') logCommand.arg = '********';

    const log = this.connection.log.child({directive: command.directive});
    log.trace({command: logCommand}, 'Handle command');

    if (!REGISTRY.hasOwnProperty(command.directive)) {
      return this.connection.reply(502, `Command not allowed: ${command.directive}`);
    }

    if (_.includes(this.blacklist, command.directive)) {
      return this.connection.reply(502, `Command blacklisted: ${command.directive}`);
    }

    if (this.whitelist.length > 0 && !_.includes(this.whitelist, command.directive)) {
      return this.connection.reply(502, `Command not whitelisted: ${command.directive}`);
    }

    const commandRegister = REGISTRY[command.directive];
    const commandFlags = _.get(commandRegister, 'flags', {});
    if (!commandFlags.no_auth && !this.connection.authenticated) {
      return this.connection.reply(530, `Command requires authentication: ${command.directive}`);
    }

    if (!commandRegister.handler) {
      return this.connection.reply(502, `Handler not set on command: ${command.directive}`);
    }

    const handler = commandRegister.handler.bind(this.connection);
    return Promise.resolve(handler({log, command, previous_command: this.previousCommand}))
    .then(() => {
      this.previousCommand = _.clone(command);
    });
  }
}
module.exports = FtpCommands;
